File: frontend/lib/services/nlp_parser.dart
import 'package:ticktick_clone/models/task.dart';

/// Result of parsing a natural language task input string.
class ParsedTaskInput {
  final String title;
  final DateTime? dueDate;
  final TaskPriority priority;
  final List<String> tags;
  final String? listName;
  final RecurrenceRule? recurrence;

  const ParsedTaskInput({
    required this.title,
    this.dueDate,
    this.priority = TaskPriority.none,
    this.tags = const [],
    this.listName,
    this.recurrence,
  });
}

/// Lightweight recurrence rule matching the backend schema.
class RecurrenceRule {
  final String frequency; // daily, weekly, monthly, yearly
  final int interval;
  final List<int>? daysOfWeek;

  const RecurrenceRule({
    required this.frequency,
    this.interval = 1,
    this.daysOfWeek,
  });

  Map<String, dynamic> toMap() => {
        'frequency': frequency,
        'interval': interval,
        'afterCompletion': false,
        if (daysOfWeek != null) 'daysOfWeek': daysOfWeek,
      };
}

// Day name → Dart weekday (DateTime.monday=1 .. DateTime.sunday=7)
// But we store 0=Sun..6=Sat to match backend
const _dayMap = <String, int>{
  'sunday': 0, 'sun': 0,
  'monday': 1, 'mon': 1,
  'tuesday': 2, 'tue': 2, 'tues': 2,
  'wednesday': 3, 'wed': 3,
  'thursday': 4, 'thu': 4, 'thurs': 4,
  'friday': 5, 'fri': 5,
  'saturday': 6, 'sat': 6,
};

/// Parse a task title string and extract structured fields.
///
/// Supported syntax:
///   Dates/times: "tomorrow", "today", "tonight", "tomorrow at 3pm",
///                "next Monday", "next week", "in 2 hours", "in 3 days",
///                "Monday", "Friday at 5pm"
///   Priority:    "!!" = high, "!" = medium
///   Tags:        "#work", "#personal"
///   List:        "/projectname"
///   Recurrence:  "every day", "every Monday", "daily", "weekly", "monthly",
///                "every 3 days", "every 2 weeks"
ParsedTaskInput parseTaskInput(String input, {DateTime? now}) {
  final refDate = now ?? DateTime.now();
  var text = input;
  DateTime? dueDate;
  var priority = TaskPriority.none;
  final tags = <String>[];
  String? listName;
  RecurrenceRule? recurrence;

  // --- Extract tags (#word) ---
  text = text.replaceAllMapped(RegExp(r'#(\w[\w-]*)'), (m) {
    tags.add(m.group(1)!);
    return '';
  });

  // --- Extract list assignment (/listname) ---
  text = text.replaceAllMapped(RegExp(r'/(\w[\w-]*)'), (m) {
    listName = m.group(1)!;
    return '';
  });

  // --- Extract priority (!! or !) ---
  if (RegExp(r'\s!!(?:\s|$)').hasMatch(text) ||
      text.startsWith('!! ') ||
      text == '!!') {
    priority = TaskPriority.high;
    text = text.replaceAll('!!', '');
  } else if (RegExp(r'\s!(?:\s|$)').hasMatch(text) ||
      text.startsWith('! ') ||
      text == '!') {
    priority = TaskPriority.medium;
    text = text.replaceAllMapped(RegExp(r'(?<![!])!(?![!])'), (_) => '');
  }

  // --- Extract recurrence ---
  final recResult = _parseRecurrence(text);
  if (recResult != null) {
    recurrence = recResult.rule;
    text = recResult.remaining;
    if (recResult.initialDay != null) {
      dueDate = _getNextDayOfWeek(refDate, recResult.initialDay!);
    }
  }

  // --- Extract date/time ---
  final dateResult = _parseDate(text, refDate);
  if (dateResult != null) {
    dueDate = dateResult.date;
    text = dateResult.remaining;
  }

  // Clean up whitespace
  final title = text.replaceAll(RegExp(r'\s+'), ' ').trim();

  return ParsedTaskInput(
    title: title,
    dueDate: dueDate,
    priority: priority,
    tags: tags,
    listName: listName,
    recurrence: recurrence,
  );
}

// --- Internal helpers ---

class _DateResult {
  final DateTime date;
  final String remaining;
  const _DateResult(this.date, this.remaining);
}

class _RecurrenceResult {
  final RecurrenceRule rule;
  final String remaining;
  final int? initialDay;
  const _RecurrenceResult(this.rule, this.remaining, [this.initialDay]);
}

_RecurrenceResult? _parseRecurrence(String text) {
  var remaining = text;

  // "every N days/weeks/months/years"
  final everyN = RegExp(
      r'\bevery\s+(\d+)\s+(day|days|week|weeks|month|months|year|years)\b',
      caseSensitive: false);
  final everyNMatch = everyN.firstMatch(remaining);
  if (everyNMatch != null) {
    final interval = int.parse(everyNMatch.group(1)!);
    final unit = everyNMatch.group(2)!.toLowerCase().replaceAll(RegExp(r's$'), '');
    const freqMap = {
      'day': 'daily',
      'week': 'weekly',
      'month': 'monthly',
      'year': 'yearly',
    };
    remaining = remaining.replaceFirst(everyN, '');
    return _RecurrenceResult(
      RecurrenceRule(frequency: freqMap[unit]!, interval: interval),
      remaining,
    );
  }

  // "every Monday", "every Monday and Wednesday"
  final everyDay = RegExp(
      r'\bevery\s+((?:(?:and|,)\s*)?(?:sun(?:day)?|mon(?:day)?|tue(?:s(?:day)?)?|wed(?:nesday)?|thu(?:rs(?:day)?)?|fri(?:day)?|sat(?:urday)?)(?:\s*(?:,|and)\s*(?:sun(?:day)?|mon(?:day)?|tue(?:s(?:day)?)?|wed(?:nesday)?|thu(?:rs(?:day)?)?|fri(?:day)?|sat(?:urday)?))*)\b',
      caseSensitive: false);
  final everyDayMatch = everyDay.firstMatch(remaining);
  if (everyDayMatch != null) {
    final days = _parseDayList(everyDayMatch.group(1)!);
    if (days.isNotEmpty) {
      remaining = remaining.replaceFirst(everyDay, '');
      return _RecurrenceResult(
        RecurrenceRule(frequency: 'weekly', interval: 1, daysOfWeek: days),
        remaining,
        days.first,
      );
    }
  }

  // "every day" / "daily"
  final daily = RegExp(r'\b(?:every\s+day|daily)\b', caseSensitive: false);
  if (daily.hasMatch(remaining)) {
    remaining = remaining.replaceFirst(daily, '');
    return _RecurrenceResult(
      const RecurrenceRule(frequency: 'daily'),
      remaining,
    );
  }

  // "weekly"
  final weekly = RegExp(r'\bweekly\b', caseSensitive: false);
  if (weekly.hasMatch(remaining)) {
    remaining = remaining.replaceFirst(weekly, '');
    return _RecurrenceResult(
      const RecurrenceRule(frequency: 'weekly'),
      remaining,
    );
  }

  // "monthly"
  final monthly = RegExp(r'\bmonthly\b', caseSensitive: false);
  if (monthly.hasMatch(remaining)) {
    remaining = remaining.replaceFirst(monthly, '');
    return _RecurrenceResult(
      const RecurrenceRule(frequency: 'monthly'),
      remaining,
    );
  }

  // "yearly"
  final yearly = RegExp(r'\byearly\b', caseSensitive: false);
  if (yearly.hasMatch(remaining)) {
    remaining = remaining.replaceFirst(yearly, '');
    return _RecurrenceResult(
      const RecurrenceRule(frequency: 'yearly'),
      remaining,
    );
  }

  return null;
}

List<int> _parseDayList(String text) {
  final days = <int>{};
  final parts = text.split(RegExp(r'\s*(?:,|and)\s*', caseSensitive: false));
  for (final part in parts) {
    final trimmed = part.trim().toLowerCase();
    if (trimmed.isNotEmpty && _dayMap.containsKey(trimmed)) {
      days.add(_dayMap[trimmed]!);
    }
  }
  return days.toList()..sort();
}

_DateResult? _parseDate(String text, DateTime refDate) {
  var remaining = text;

  // "today" / "today at HH:MM am/pm"
  final today = RegExp(
      r'\btoday(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?',
      caseSensitive: false);
  final todayMatch = today.firstMatch(remaining);
  if (todayMatch != null) {
    var date = DateTime(refDate.year, refDate.month, refDate.day, 23, 59);
    if (todayMatch.group(1) != null) {
      date = _applyTime(
          date, todayMatch.group(1)!, todayMatch.group(2), todayMatch.group(3));
    }
    remaining = remaining.replaceFirst(today, '');
    return _DateResult(date, remaining);
  }

  // "tonight"
  final tonight = RegExp(r'\btonight\b', caseSensitive: false);
  if (tonight.hasMatch(remaining)) {
    final date = DateTime(refDate.year, refDate.month, refDate.day, 21, 0);
    remaining = remaining.replaceFirst(tonight, '');
    return _DateResult(date, remaining);
  }

  // "tomorrow" / "tomorrow at Hpm"
  final tomorrow = RegExp(
      r'\btomorrow(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?',
      caseSensitive: false);
  final tomorrowMatch = tomorrow.firstMatch(remaining);
  if (tomorrowMatch != null) {
    final nextDay = refDate.add(const Duration(days: 1));
    var date = DateTime(nextDay.year, nextDay.month, nextDay.day, 23, 59);
    if (tomorrowMatch.group(1) != null) {
      date = _applyTime(date, tomorrowMatch.group(1)!, tomorrowMatch.group(2),
          tomorrowMatch.group(3));
    }
    remaining = remaining.replaceFirst(tomorrow, '');
    return _DateResult(date, remaining);
  }

  // "next week"
  final nextWeek = RegExp(r'\bnext\s+week\b', caseSensitive: false);
  if (nextWeek.hasMatch(remaining)) {
    final date = _getNextDayOfWeek(refDate, 1); // next Monday
    final dateWithTime = DateTime(date.year, date.month, date.day, 9, 0);
    remaining = remaining.replaceFirst(nextWeek, '');
    return _DateResult(dateWithTime, remaining);
  }

  // "next Monday", "next Friday at 5pm"
  final nextDay = RegExp(
      r'\bnext\s+(sunday|sun|monday|mon|tuesday|tue|tues|wednesday|wed|thursday|thu|thurs|friday|fri|saturday|sat)(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?\b',
      caseSensitive: false);
  final nextDayMatch = nextDay.firstMatch(remaining);
  if (nextDayMatch != null) {
    final dayName = nextDayMatch.group(1)!.toLowerCase();
    final targetDay = _dayMap[dayName]!;
    final date = _getNextDayOfWeek(refDate, targetDay, forceNextWeek: true);
    var dateWithTime = DateTime(date.year, date.month, date.day, 9, 0);
    if (nextDayMatch.group(2) != null) {
      dateWithTime = _applyTime(dateWithTime, nextDayMatch.group(2)!,
          nextDayMatch.group(3), nextDayMatch.group(4));
    }
    remaining = remaining.replaceFirst(nextDay, '');
    return _DateResult(dateWithTime, remaining);
  }

  // "in N hours/minutes/days/weeks"
  final inN = RegExp(
      r'\bin\s+(\d+)\s+(hour|hours|minute|minutes|min|mins|day|days|week|weeks)\b',
      caseSensitive: false);
  final inNMatch = inN.firstMatch(remaining);
  if (inNMatch != null) {
    final n = int.parse(inNMatch.group(1)!);
    final unit = inNMatch.group(2)!.toLowerCase().replaceAll(RegExp(r's$'), '');
    DateTime date;
    switch (unit) {
      case 'hour':
        date = refDate.add(Duration(hours: n));
      case 'minute':
      case 'min':
        date = refDate.add(Duration(minutes: n));
      case 'day':
        date = refDate.add(Duration(days: n));
      case 'week':
        date = refDate.add(Duration(days: n * 7));
      default:
        date = refDate;
    }
    remaining = remaining.replaceFirst(inN, '');
    return _DateResult(date, remaining);
  }

  // Bare day name: "Monday", "Friday at 5pm"
  final bareDay = RegExp(
      r'\b(sunday|sun|monday|mon|tuesday|tue|tues|wednesday|wed|thursday|thu|thurs|friday|fri|saturday|sat)(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?\b',
      caseSensitive: false);
  final bareDayMatch = bareDay.firstMatch(remaining);
  if (bareDayMatch != null) {
    final dayName = bareDayMatch.group(1)!.toLowerCase();
    final targetDay = _dayMap[dayName]!;
    final date = _getNextDayOfWeek(refDate, targetDay);
    var dateWithTime = DateTime(date.year, date.month, date.day, 9, 0);
    if (bareDayMatch.group(2) != null) {
      dateWithTime = _applyTime(dateWithTime, bareDayMatch.group(2)!,
          bareDayMatch.group(3), bareDayMatch.group(4));
    }
    remaining = remaining.replaceFirst(bareDay, '');
    return _DateResult(dateWithTime, remaining);
  }

  return null;
}

DateTime _applyTime(
    DateTime date, String hourStr, String? minuteStr, String? ampm) {
  var hour = int.parse(hourStr);
  final minute = minuteStr != null ? int.parse(minuteStr) : 0;
  if (ampm != null) {
    final isPm = ampm.toLowerCase() == 'pm';
    if (isPm && hour < 12) hour += 12;
    if (!isPm && hour == 12) hour = 0;
  } else if (hour < 8) {
    // Assume PM for ambiguous small numbers
    hour += 12;
  }
  return DateTime(date.year, date.month, date.day, hour, minute);
}

DateTime _getNextDayOfWeek(DateTime refDate, int targetDay,
    {bool forceNextWeek = false}) {
  // Convert from 0=Sun..6=Sat to Dart's 1=Mon..7=Sun
  final dartTarget = targetDay == 0 ? 7 : targetDay;
  final currentDay = refDate.weekday; // 1=Mon..7=Sun
  var daysAhead = dartTarget - currentDay;
  if (daysAhead <= 0 || forceNextWeek) {
    daysAhead += 7;
  }
  return refDate.add(Duration(days: daysAhead));
}

File: backend/src/services/nlp-parser.ts
import { Priority, RecurrenceRule, RecurrenceFrequency } from "../models/schemas";

export interface ParsedTask {
  title: string; // cleaned title with parsed tokens removed
  dueDate?: string; // ISO datetime
  priority: Priority;
  tags: string[];
  listName?: string; // raw name (caller resolves to listId)
  recurrence?: RecurrenceRule;
}

// Day name → JS day number (0=Sun..6=Sat)
const DAY_MAP: Record<string, number> = {
  sunday: 0,
  sun: 0,
  monday: 1,
  mon: 1,
  tuesday: 2,
  tue: 2,
  tues: 2,
  wednesday: 3,
  wed: 3,
  thursday: 4,
  thu: 4,
  thurs: 4,
  friday: 5,
  fri: 5,
  saturday: 6,
  sat: 6,
};

/**
 * Parse a task title string and extract structured fields.
 *
 * Supported syntax:
 *   Dates/times: "tomorrow", "today", "tonight", "tomorrow at 3pm",
 *                "next Monday", "next week", "in 2 hours", "in 3 days",
 *                "Monday", "Friday at 5pm"
 *   Priority:    "!!" = high, "!" = medium
 *   Tags:        "#work", "#personal"
 *   List:        "/projectname"
 *   Recurrence:  "every day", "every Monday", "daily", "weekly", "monthly",
 *                "every 3 days", "every 2 weeks"
 *
 * @param input  raw task title text
 * @param now    optional reference date for relative calculations (defaults to Date.now())
 * @returns ParsedTask with cleaned title and extracted fields
 */
export function parseTaskInput(input: string, now?: Date): ParsedTask {
  const refDate = now ?? new Date();
  let text = input;
  let dueDate: Date | undefined;
  let priority: Priority = "none";
  const tags: string[] = [];
  let listName: string | undefined;
  let recurrence: RecurrenceRule | undefined;

  // --- Extract tags (#word) ---
  text = text.replace(/#(\w[\w-]*)/g, (_match, tag: string) => {
    tags.push(tag);
    return "";
  });

  // --- Extract list assignment (/listname) ---
  text = text.replace(/\/(\w[\w-]*)/g, (_match, name: string) => {
    listName = name;
    return "";
  });

  // --- Extract priority (!! or !) ---
  // Must check !! before ! to avoid partial match
  if (/\s!!(?:\s|$)/.test(text) || text.startsWith("!! ") || text === "!!") {
    priority = "high";
    text = text.replace(/!!/g, "");
  } else if (/\s!(?:\s|$)/.test(text) || text.startsWith("! ") || text === "!") {
    priority = "medium";
    text = text.replace(/(?<![!])!(?![!])/g, "");
  }

  // --- Extract recurrence (must be before date parsing since "every Monday" contains a day) ---
  const recurrenceResult = parseRecurrence(text);
  if (recurrenceResult) {
    recurrence = recurrenceResult.rule;
    text = recurrenceResult.remaining;
    // If recurrence specifies a day, set the initial due date to the next occurrence
    if (recurrenceResult.initialDay !== undefined && !dueDate) {
      dueDate = getNextDayOfWeek(refDate, recurrenceResult.initialDay);
    }
  }

  // --- Extract date/time ---
  const dateResult = parseDate(text, refDate);
  if (dateResult) {
    dueDate = dateResult.date;
    text = dateResult.remaining;
  }

  // Clean up extra whitespace
  const title = text.replace(/\s+/g, " ").trim();

  const result: ParsedTask = { title, priority, tags };
  if (dueDate) {
    result.dueDate = dueDate.toISOString();
  }
  if (listName) {
    result.listName = listName;
  }
  if (recurrence) {
    result.recurrence = recurrence;
  }
  return result;
}

interface DateParseResult {
  date: Date;
  remaining: string;
}

interface RecurrenceParseResult {
  rule: RecurrenceRule;
  remaining: string;
  initialDay?: number; // day of week for initial due date
}

function parseRecurrence(text: string): RecurrenceParseResult | null {
  let remaining = text;

  // "every N days/weeks/months/years"
  const everyNPattern =
    /\bevery\s+(\d+)\s+(day|days|week|weeks|month|months|year|years)\b/i;
  const everyNMatch = remaining.match(everyNPattern);
  if (everyNMatch) {
    const interval = parseInt(everyNMatch[1], 10);
    const unit = everyNMatch[2].toLowerCase().replace(/s$/, "");
    const freqMap: Record<string, RecurrenceFrequency> = {
      day: "daily",
      week: "weekly",
      month: "monthly",
      year: "yearly",
    };
    remaining = remaining.replace(everyNPattern, "");
    return {
      rule: { frequency: freqMap[unit], interval, afterCompletion: false },
      remaining,
    };
  }

  // "every Monday", "every Tuesday and Thursday", etc.
  const everyDayPattern = /\bevery\s+((?:(?:and|,)\s*)?(?:sun(?:day)?|mon(?:day)?|tue(?:s(?:day)?)?|wed(?:nesday)?|thu(?:rs(?:day)?)?|fri(?:day)?|sat(?:urday)?)(?:\s*(?:,|and)\s*(?:sun(?:day)?|mon(?:day)?|tue(?:s(?:day)?)?|wed(?:nesday)?|thu(?:rs(?:day)?)?|fri(?:day)?|sat(?:urday)?))*)\b/i;
  const everyDayMatch = remaining.match(everyDayPattern);
  if (everyDayMatch) {
    const dayStr = everyDayMatch[1];
    const days = parseDayList(dayStr);
    if (days.length > 0) {
      remaining = remaining.replace(everyDayPattern, "");
      return {
        rule: { frequency: "weekly", interval: 1, daysOfWeek: days, afterCompletion: false },
        remaining,
        initialDay: days[0],
      };
    }
  }

  // "every day" / "daily"
  const dailyPattern = /\b(?:every\s+day|daily)\b/i;
  if (dailyPattern.test(remaining)) {
    remaining = remaining.replace(dailyPattern, "");
    return { rule: { frequency: "daily", interval: 1, afterCompletion: false }, remaining };
  }

  // "weekly"
  const weeklyPattern = /\bweekly\b/i;
  if (weeklyPattern.test(remaining)) {
    remaining = remaining.replace(weeklyPattern, "");
    return { rule: { frequency: "weekly", interval: 1, afterCompletion: false }, remaining };
  }

  // "monthly"
  const monthlyPattern = /\bmonthly\b/i;
  if (monthlyPattern.test(remaining)) {
    remaining = remaining.replace(monthlyPattern, "");
    return { rule: { frequency: "monthly", interval: 1, afterCompletion: false }, remaining };
  }

  // "yearly"
  const yearlyPattern = /\byearly\b/i;
  if (yearlyPattern.test(remaining)) {
    remaining = remaining.replace(yearlyPattern, "");
    return { rule: { frequency: "yearly", interval: 1, afterCompletion: false }, remaining };
  }

  return null;
}

function parseDayList(text: string): number[] {
  const days: number[] = [];
  // Split on comma, "and", or whitespace clusters
  const parts = text.split(/\s*(?:,|and)\s*/i);
  for (const part of parts) {
    const trimmed = part.trim().toLowerCase();
    if (trimmed && DAY_MAP[trimmed] !== undefined) {
      days.push(DAY_MAP[trimmed]);
    }
  }
  return [...new Set(days)].sort((a, b) => a - b);
}

function parseDate(text: string, refDate: Date): DateParseResult | null {
  let remaining = text;

  // "today" / "today at HH:MM" / "today at Hpm"
  const todayPattern = /\btoday(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?/i;
  const todayMatch = remaining.match(todayPattern);
  if (todayMatch) {
    const date = new Date(refDate);
    if (todayMatch[1]) {
      applyTime(date, todayMatch[1], todayMatch[2], todayMatch[3]);
    } else {
      date.setHours(23, 59, 0, 0);
    }
    remaining = remaining.replace(todayPattern, "");
    return { date, remaining };
  }

  // "tonight"
  const tonightPattern = /\btonight\b/i;
  const tonightMatch = remaining.match(tonightPattern);
  if (tonightMatch) {
    const date = new Date(refDate);
    date.setHours(21, 0, 0, 0);
    remaining = remaining.replace(tonightPattern, "");
    return { date, remaining };
  }

  // "tomorrow" / "tomorrow at 3pm"
  const tomorrowPattern =
    /\btomorrow(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?/i;
  const tomorrowMatch = remaining.match(tomorrowPattern);
  if (tomorrowMatch) {
    const date = new Date(refDate);
    date.setDate(date.getDate() + 1);
    if (tomorrowMatch[1]) {
      applyTime(date, tomorrowMatch[1], tomorrowMatch[2], tomorrowMatch[3]);
    } else {
      date.setHours(23, 59, 0, 0);
    }
    remaining = remaining.replace(tomorrowPattern, "");
    return { date, remaining };
  }

  // "next week" (next Monday)
  const nextWeekPattern = /\bnext\s+week\b/i;
  const nextWeekMatch = remaining.match(nextWeekPattern);
  if (nextWeekMatch) {
    const date = getNextDayOfWeek(refDate, 1); // next Monday
    date.setHours(9, 0, 0, 0);
    remaining = remaining.replace(nextWeekPattern, "");
    return { date, remaining };
  }

  // "next Monday", "next Friday at 5pm"
  const nextDayPattern =
    /\bnext\s+(sunday|sun|monday|mon|tuesday|tue|tues|wednesday|wed|thursday|thu|thurs|friday|fri|saturday|sat)(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?\b/i;
  const nextDayMatch = remaining.match(nextDayPattern);
  if (nextDayMatch) {
    const dayName = nextDayMatch[1].toLowerCase();
    const targetDay = DAY_MAP[dayName];
    const date = getNextDayOfWeek(refDate, targetDay, true); // force next week
    if (nextDayMatch[2]) {
      applyTime(date, nextDayMatch[2], nextDayMatch[3], nextDayMatch[4]);
    } else {
      date.setHours(9, 0, 0, 0);
    }
    remaining = remaining.replace(nextDayPattern, "");
    return { date, remaining };
  }

  // "in N hours/minutes/days/weeks"
  const inNPattern =
    /\bin\s+(\d+)\s+(hour|hours|minute|minutes|min|mins|day|days|week|weeks)\b/i;
  const inNMatch = remaining.match(inNPattern);
  if (inNMatch) {
    const n = parseInt(inNMatch[1], 10);
    const unit = inNMatch[2].toLowerCase().replace(/s$/, "");
    const date = new Date(refDate);
    switch (unit) {
      case "hour":
        date.setHours(date.getHours() + n);
        break;
      case "minute":
      case "min":
        date.setMinutes(date.getMinutes() + n);
        break;
      case "day":
        date.setDate(date.getDate() + n);
        break;
      case "week":
        date.setDate(date.getDate() + n * 7);
        break;
    }
    remaining = remaining.replace(inNPattern, "");
    return { date, remaining };
  }

  // Bare day name: "Monday", "Friday at 5pm" (meaning this coming occurrence)
  const bareDayPattern =
    /\b(sunday|sun|monday|mon|tuesday|tue|tues|wednesday|wed|thursday|thu|thurs|friday|fri|saturday|sat)(?:\s+at\s+(\d{1,2})(?::(\d{2}))?\s*(am|pm)?)?\b/i;
  const bareDayMatch = remaining.match(bareDayPattern);
  if (bareDayMatch) {
    const dayName = bareDayMatch[1].toLowerCase();
    const targetDay = DAY_MAP[dayName];
    // Only match if the day name stands on its own (not part of another word)
    const date = getNextDayOfWeek(refDate, targetDay);
    if (bareDayMatch[2]) {
      applyTime(date, bareDayMatch[2], bareDayMatch[3], bareDayMatch[4]);
    } else {
      date.setHours(9, 0, 0, 0);
    }
    remaining = remaining.replace(bareDayPattern, "");
    return { date, remaining };
  }

  return null;
}

function applyTime(
  date: Date,
  hourStr: string,
  minuteStr?: string,
  ampm?: string
): void {
  let hour = parseInt(hourStr, 10);
  const minute = minuteStr ? parseInt(minuteStr, 10) : 0;
  if (ampm) {
    const isPm = ampm.toLowerCase() === "pm";
    if (isPm && hour < 12) hour += 12;
    if (!isPm && hour === 12) hour = 0;
  } else if (hour < 8) {
    // Assume PM for small numbers without explicit am/pm (e.g., "at 3" → 3pm)
    hour += 12;
  }
  date.setHours(hour, minute, 0, 0);
}

function getNextDayOfWeek(
  refDate: Date,
  targetDay: number,
  forceNextWeek = false
): Date {
  const date = new Date(refDate);
  const currentDay = date.getDay();
  let daysAhead = targetDay - currentDay;
  if (daysAhead <= 0 || forceNextWeek) {
    daysAhead += 7;
  }
  if (forceNextWeek && daysAhead <= 7) {
    // "next Monday" when today is Sunday should go to Monday after this week
    // If daysAhead is exactly 7, that's fine (next week's same day)
  }
  date.setDate(date.getDate() + daysAhead);
  return date;
}
